README.md
![BitReader logo](https://github.com/end2endzone/BitReader/raw/master/docs/BitReader-splashscreen.png)


[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![Github Releases](https://img.shields.io/github/release/end2endzone/BitReader.svg)](https://github.com/end2endzone/BitReader/releases)



# BitReader #

The BitReader library is an arduino library that allows one to read or write data which is not aligned on 8, 16 or 32 bits variables. In other words, the library allows one to quickly store (or read) multiple variables from binary buffers.

Library features:
* Reduces size of data (strings, structures, arrays) in memory to the minimum amount of bits required.
* Help reading/updating bit-field structures.
* Handles bit [serialization](http://en.wikipedia.org/wiki/Serialization) data to/from buffers.



## Status ##

Build:

| Service | Build | Tests |
|----|-------|-------|
| AppVeyor | [![Build status](https://img.shields.io/appveyor/ci/end2endzone/BitReader/master.svg?logo=appveyor)](https://ci.appveyor.com/project/end2endzone/BitReader) | [![Tests status](https://img.shields.io/appveyor/tests/end2endzone/BitReader/master.svg?logo=appveyor)](https://ci.appveyor.com/project/end2endzone/BitReader/branch/master/tests) |
| Travis CI | [![Build Status](https://img.shields.io/travis/end2endzone/BitReader/master.svg?logo=travis&style=flat)](https://travis-ci.org/end2endzone/BitReader) |  |

Statistics:

| AppVeyor | Travic CI |
|----------|-----------|
| [![Statistics](https://buildstats.info/appveyor/chart/end2endzone/BitReader)](https://ci.appveyor.com/project/end2endzone/BitReader/branch/master) | [![Statistics](https://buildstats.info/travisci/chart/end2endzone/BitReader)](https://travis-ci.org/end2endzone/BitReader) |



# Purpose #

Desktop application developers are not used to have memory (or even CPU) requirements. Most will not look at memory consumption as an arduino developer would. Memory & CPU is cheap. If an application is struggling to run on today's computer, it will run smoothly in 1-2 years. Arduino developers know that the amount of memory available on embedded devices is pretty limited. Having a way to save bits of precious memory is more than interesting.

The conventional way of storing bit data into sequential buffers is to use a combination of [bit mask](http://en.wikipedia.org/wiki/Mask_(computing)) and [bit shift](http://en.wikipedia.org/wiki/Logical_shift):

```cpp
//define storage for multiple values
uint8_t value1; //5 bits
uint8_t value2; //8 bits
uint8_t value3; //1 bit
uint8_t value4; //6 bits

uint8_t buffer[] = {0xAA, 0xAA, 0x0F};

// buffer[2] buffer[1] buffer[0]
//  00001111  10101010  10101010
//                         |---| ==> value1
//               |----  --| ==> value2
//              | ==> value3
//      |---  -| ==> value4

value1 =   buffer[0] & 0x1F;        //first 5 bits of buffer[0]
value2 = ((buffer[0] & 0xE0) >> 5)+ //next  3 bits of buffer[0]
         ((buffer[1] & 0x1F) << 3); //first 5 bits of buffer[1]
value3 = ((buffer[1] & 0x20) >> 5); //next    bit  of buffer[1]
value4 = ((buffer[1] & 0xC0) >> 6)+ //next  2 bits of buffer[1]
         ((buffer[2] & 0x0F) << 2); //next  4 bits of buffer[2]

//expect value1 == 0x0A
//expect value2 == 0x55
//expect value3 == 0x01
//expect value4 == 0x3E
```

This approach works great if you do not make a change on existing code but it gets pretty complicated and error prone as soon as you need to make a change:

* Insert `value0` (2 bits) at the beginning.
* Insert a new value between `value3` and `value4`.
* Change the size of `value3`
* etc...

In these kind of situations, the BitReader library becomes handy and takes care of all the bit handling complexity.



# Usage #

The following instructions show how to use the library.




## General ##

Create an instance of `BitReader`, `BitWriter` or `BitAddress` depending on the following optimization switches:
* USE_BITADDRESS_SETTER_GETTER
* USE_SINGLEBIT_MACROS
* USE_BITADDRESS_READ_WRITE

Then assign a reading or writing buffer using the `setBuffer()` method.

Call the `write()` or `read()` methods for writing or reading bits to the assigned buffer. Each method allows one to specify the amount of bits to read or write.




## Making text strings shorter ##

The library is useful for storing or decoding data in binary format using the **minimum amount of bits**.

Storing words composed of only lower case letters, spaces and dots requires only 5 bits per characters instead of 8 bits (which saves 3 bits per character). The phase `hello my name is antoine. i wrote the bitreader library.` takes 56 bytes as an array of char but uses 35 bytes as 5 bits chunks.

Allow 6 bits per characters and you can also include capital letters and numbers: `Hello my name is Antoine. I wrote the BitReader library when I was 34 years old.` takes 80 bytes as an array of char but only requires 60 bytes as 6 bits chunks.




## Shorter structures definition ##

The library is particularly useful for dealing with [bit-field structures](http://en.cppreference.com/w/cpp/language/bit_field).

Consider the following:

```cpp
struct Person {
  char name[10];
  char birthDay;
  char birthMonth;
  short birthYear;
  bool maried;
  char children;
};
```

The `Person` structure requires 16 bytes in memory (10+1+1+2+1+1) or 128 bits. However, using bit-field structure, the `Person` structure can be defined as the following:

```cpp
struct Person2 {
 uint8_t name[10]; //keep 8 bits per letters for compatibility.
 union
 {
   struct //aligned on 8 bits
   {
     uint8_t birthDay : 5; //from 1 to 31
     uint8_t children : 3; //from 0 to 7. 
     uint8_t birthMonth : 4; //from 1 to 12
     uint8_t maried : 1; //true or false
     //uint8_t : 3; //birthYear (low)
     //uint8_t : 8; //birthYear (high)
   };
   struct //aligned on 32 bits
   {
     uint32_t padding1 :13; //padding for birthDay, children, birthMonth, maried
     uint32_t birthYear :12; //allows years from 0 to 4095.
     //uint16_t : 4; //unused
   };
 };
};
```

The `Person2` structure contains the same information as the `Person` structure but instead of using 128 bits in memory it only requires 105 bits (80+5+3+4+1+12).

A person could also be defined with masks:

```cpp
struct Person3 {
  uint8_t name[10];
  uint32_t buffer;
};
static const uint32_t MASK_BIRTHDAY = (1<<4 | 1<<3 | 1<<2 | 1<<1 | 1<<0);
static const uint32_t MASK_CHILDREN = (1<<7 | 1<<6 | 1<<5);
static const uint32_t MASK_BIRTHMONTH = (1<<11 | 1<<10 | 1<<9 | 1<<8);
static const uint32_t MASK_MARIED = (1<<12);
static const uint32_t MASK_BIRTHYEAR =  (1<<24 | 1<<23 | 1<<22 | 1<<21 | 1<<20 | 1<<19 | 1<<18 | 1<<17 | 1<<16 | 1<<15 | 1<<14 | 1<<13);
static const uint32_t MASK_PADDING = (1<<31 | 1<<30 | 1<<29 | 1<<28 | 1<<27 | 1<<26 | 1<<25);
```

In the last 2 scenarios (`Person2` and `Person3`), the BitReader library allows one to decode a `Person` structure from a binary buffer.




## Make data arrays much shorter ##

Consider an algorithm that plays a [Morse code](https://en.wikipedia.org/wiki/Morse_code). Morse code defines 3 symbols that can be played: dots, dashes and pauses.

For example, the following string `Hello my name is Antoine. I wrote the BitReader library when I was 34 years old.` (56 bytes) translate into Morse code as:

```
.... . .-.. .-.. --- / -- -.-- / -. .- -- . / .. ... / .- -. 
- --- .. -. . .-.-.- / .. / .-- .-. --- - . / - .... . / -... 
.. - .-. . .- -.. . .-. / .-.. .. -... .-. .- .-. -.-- / .-- 
.... . -. / .. / .-- .- ... / ...-- ....- / -.-- . .- .-. 
... / --- .-.. -.. .-.-.-
```

According to [this translator](http://morsecode.scphillips.com/translator.html). The whole code takes 267 bytes in memory. However, using 2 bits per code, the whole string can be encoded in a char buffer with only 534 bits (~67 bytes).

The same concept applies to all numeric array.




## Sample code ##

The following demo show how to use the library:

```cpp
#include <bitreader.h>

/**
* Encoding "text" as 5 bits characters using the following:
*   A -> 0
*   Z -> 25
*     -> 26
*   . -> 27
*   0 -> 28
*   1 -> 29
*   2 -> 30
*   3 -> 31
*/

#define BUFFER_SIZE 600
unsigned char sourceBuffer[BUFFER_SIZE];
unsigned char targetBuffer[BUFFER_SIZE];

void setup() {
  Serial.begin(9600);

  #ifdef USE_BITADDRESS_SETTER_GETTER
  Serial.println("#define USE_BITADDRESS_SETTER_GETTER");
  #endif
  #ifdef USE_SINGLEBIT_MACROS
  Serial.println("#define USE_SINGLEBIT_MACROS");
  #endif
  #ifdef USE_BITADDRESS_READ_WRITE
  Serial.println("#define USE_BITADDRESS_READ_WRITE");
  #endif

  //copy all characters from sourceBuffer to targetBuffer using 1 to 16 bits block.
  for(int numBits = 1; numBits<=16; numBits++) {
    #ifndef USE_BITADDRESS_READ_WRITE
    BitReader bitreader;
    BitWriter bitwriter;
    #else
    BitAddress bitreader;
    BitAddress bitwriter;
    #endif
    
    bitreader.setBuffer(sourceBuffer);
    bitwriter.setBuffer(targetBuffer);

    //initialize sourceBuffer
    for(int i=0; i<BUFFER_SIZE; i++) {
      sourceBuffer[i] = i+numBits;
    }

    //start counter
    unsigned long timeStart = millis();

    //proceed with buffer copy
    int totalBits = BUFFER_SIZE*BITS_PER_BYTE;
    int numBlocks = totalBits / numBits;
    int temp = 0;
    for(int i=0; i<numBlocks; i++) {
      bitreader.read(numBits, &temp);
      bitwriter.write(&temp, numBits);
    }

    //stop timer
    unsigned long timeEnd = millis();
    unsigned long timeElapsed = timeEnd - timeStart;

    //make sure both buffer are equals (that bit copy operation worked as expected)
    bool equal = true;
    for(int i=0; i<BUFFER_SIZE; i++) {

      // Do not validate the last 2 bytes of the buffer
      // With certain numBits size, the remaining bits
      // of the source buffer are not copied to the 
      // target buffer. The same as would 3, 6 and 7 bits
      // would not copy the totality of a byte (8 bits)
      if (i >= BUFFER_SIZE-2) {
        continue;
      }
      
      if (sourceBuffer[i] != targetBuffer[i]) {
        equal = false;
      }
    }
    
    //outputting benchnmark test values
    if (!equal) {
      Serial.println("*** Warning! Both buffers are not equals!");
    }
    char msg[65];
    sprintf(msg, "Copied %d bytes using blocks of %d bits in %d milliseconds.", BUFFER_SIZE, numBits, timeElapsed);
    Serial.println(msg);
  }
}

void loop() {
}
```




## Other samples ##

See the library examples for more details:

* [Benchmark.ino](/src/BitReader/examples/Benchmark/Benchmark.ino)
* [ReadBits.ino](src/BitReader/examples/ReadBits/ReadBits.ino)
* [WriteBits.ino](src/BitReader/examples/WriteBits/WriteBits.ino)




# Building #

Please refer to file [INSTALL.md](INSTALL.md) for details on how installing/building the application.




# Platforms #

win32Arduino has been tested with the following platform:

  * Linux x86/x64
  * Windows x86/x64




# Versioning #

We use [Semantic Versioning 2.0.0](http://semver.org/) for versioning. For the versions available, see the [tags on this repository](https://github.com/end2endzone/BitReader/tags).




# Authors #

* **Antoine Beauchamp** - *Initial work* - [end2endzone](https://github.com/end2endzone)

See also the list of [contributors](https://github.com/end2endzone/BitReader/blob/master/AUTHORS) who participated in this project.




# License #

This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details


CHANGES
Changes for 1.3.0:

* Fixed issue #2 - Create INSTALL.md file.
* Fixed issue #4 - Update repository structure to Arduino Library Format 1.5.
* Fixed issue #5 - Upgrade to win32Arduino version 2.3.0-rc.1 for executing unit tests.

Changes for 1.2.1:

* Fixed issue #1 - Update all examples: Z is character 25, not 26
* Fixed issue #3 - Use correct field separator in keywords.txt

Changes for 1.2.0:

* Source code now depends on win32Arduino third party library available at http://github.com/end2endzone/win32Arduino
* First github release.
* Code originally release at http://www.end2endzone.com/bitreader-a-collection-of-software-timers-to-easily-compute-elapsed-time-since-an-event-occured/

Changes for 1.1.0:

* Optimized bit operations

Changes for 1.0.0:
* Initial release.


test/main.cpp
#include <stdio.h>
#include <iostream>

#include <gtest/gtest.h>

int main(int argc, char **argv)
{
  //define default values for xml output report
#ifdef NDEBUG
  ::testing::GTEST_FLAG(output) = "xml:bitreader_unittest.release.xml";
#else
  ::testing::GTEST_FLAG(output) = "xml:bitreader_unittest.debug.xml";
#endif

  ::testing::GTEST_FLAG(filter) = "*";
  ::testing::InitGoogleTest(&argc, argv);

  int wResult = RUN_ALL_TESTS(); //Find and run all tests
  return wResult; // returns 0 if all the tests are successful, or 1 otherwise
}


test/TestBitReader.cpp
#include <stdio.h>
#include <string.h>
#include <string>

#include "TestBitReader.h"
#include "Arduino.h"
#include "bitreader.h"
#include "rapidassist/gtesthelp.h"

using namespace testarduino;

namespace arduino { namespace test
{
  static const size_t PERSON_NAME_SIZE = 10;

  struct Person {
    char name[PERSON_NAME_SIZE];  //10*7 bits, 70 bits
    uint8_t birthDay;             //from 1 to 31, 5 bits
    uint8_t birthMonth;           //from 1 to 12, 4 bits
    uint16_t birthYear;           //from 1900 to 2149, 250 years, 8 bits
    bool maried;                  //1 bit
    uint8_t children;             //from 0 to 7, 3 bits
  };

  //--------------------------------------------------------------------------------------------------
  void TestBitReader::SetUp()
  {
    //disable function logging
    testarduino::setLogFile("");
  }
  //--------------------------------------------------------------------------------------------------
  void TestBitReader::TearDown()
  {
  }
  //--------------------------------------------------------------------------------------------------
  TEST_F(TestBitReader, testToAbsBitOffset)
  {
    #ifndef USE_BITADDRESS_READ_WRITE
    BitWriter bitwriter;
    #else
    BitAddress bitwriter;
    #endif

    //define an output buffer
    static const size_t MAX_BUFFER_SIZE = 30;
    uint8_t buffer[MAX_BUFFER_SIZE] = {0};

    //setup BitWriter
    bitwriter.setBuffer(buffer);

    //act & assert
    ASSERT_EQ(0, bitwriter.toAbsBitOffset());
    bitwriter.add(4);
    ASSERT_EQ(4, bitwriter.toAbsBitOffset());
    bitwriter.add(1);
    ASSERT_EQ(5, bitwriter.toAbsBitOffset());
    bitwriter.add(1);
    ASSERT_EQ(6, bitwriter.toAbsBitOffset());
    bitwriter.add(1);
    ASSERT_EQ(7, bitwriter.toAbsBitOffset());
    bitwriter.add(1); //1 more byte
    ASSERT_EQ(8, bitwriter.toAbsBitOffset());
    bitwriter.add(1);
    ASSERT_EQ(9, bitwriter.toAbsBitOffset());
    bitwriter.add(1);
    ASSERT_EQ(10, bitwriter.toAbsBitOffset());
  }
  //--------------------------------------------------------------------------------------------------
  TEST_F(TestBitReader, testReset)
  {
    #ifndef USE_BITADDRESS_READ_WRITE
    BitWriter bitwriter;
    #else
    BitAddress bitwriter;
    #endif

    //define an input buffer
    static const uint8_t INPUT_BUFFER[] = {0xaa, 0xbb, 0xcc, 0xdd, 0xee, 0xff};
    static const int INPUT_BUFFER_SIZE = sizeof(INPUT_BUFFER)/sizeof(INPUT_BUFFER[0]);

    //define an output buffer
    static const size_t MAX_BUFFER_SIZE = 30;
    uint8_t buffer[MAX_BUFFER_SIZE] = {0};

    //setup BitWriter
    bitwriter.setBuffer(buffer);

    //serialize, assert
    ASSERT_EQ(0, bitwriter.toAbsBitOffset());
    bitwriter.write(&INPUT_BUFFER[0], 8);
    bitwriter.write(&INPUT_BUFFER[1], 8);
    bitwriter.reset();
    ASSERT_EQ(0, bitwriter.toAbsBitOffset());
    bitwriter.write(&INPUT_BUFFER[2], 8);
    bitwriter.write(&INPUT_BUFFER[3], 8);
    bitwriter.write(&INPUT_BUFFER[4], 8);
    bitwriter.write(&INPUT_BUFFER[5], 8);

    //compute total size of serialized data
    uint32_t numBits = bitwriter.toAbsBitOffset();
    uint32_t numBytes = numBits/8;
    if (numBits%8)
      numBytes++;

    //define ecpected buffer
    static const uint8_t EXPECTED_BUFFER[] = {0xcc, 0xdd, 0xee, 0xff}; //expected serialized buffer content
    static const int EXPECTED_BUFFER_SIZE = sizeof(EXPECTED_BUFFER)/sizeof(EXPECTED_BUFFER[0]);

    //assert size of serialized buffer
    ASSERT_EQ(32, numBits);
    ASSERT_EQ(EXPECTED_BUFFER_SIZE, numBytes);

    //assert serialized buffer content
    for(int i=0; i<EXPECTED_BUFFER_SIZE; i++)
    {
      const uint8_t EXPECTED_CHARACTER = EXPECTED_BUFFER[i];
      const uint8_t actualCharacter = buffer[i];
      ASSERT_EQ(EXPECTED_CHARACTER, actualCharacter);
    }
  }
  //--------------------------------------------------------------------------------------------------
  TEST_F(TestBitReader, testAdd8bits)
  {
    #ifndef USE_BITADDRESS_READ_WRITE
    BitWriter bitwriter;
    #else
    BitAddress bitwriter;
    #endif

    //define an input buffer
    static const uint8_t INPUT_BUFFER[] = {0xaa, 0xbb, 0xcc};
    static const int INPUT_BUFFER_SIZE = sizeof(INPUT_BUFFER)/sizeof(INPUT_BUFFER[0]);

    //define an output buffer
    static const size_t MAX_BUFFER_SIZE = 30;
    uint8_t buffer[MAX_BUFFER_SIZE] = {0};

    //setup BitWriter
    bitwriter.setBuffer(buffer);

    //serialize
    bitwriter.add(4);
    for(int i=0; i<INPUT_BUFFER_SIZE; i++)
    {
      bitwriter.write(&INPUT_BUFFER[i], 8);
    }

    //compute total size of serialized data
    uint32_t numBits = bitwriter.toAbsBitOffset();
    uint32_t numBytes = numBits/8;
    if (numBits%8)
      numBytes++;

    //define ecpected buffer
    static const uint8_t EXPECTED_BUFFER[] = {0xa0, 0xba, 0xcb, 0x0c}; //expected serialized buffer content
    static const int EXPECTED_BUFFER_SIZE = sizeof(EXPECTED_BUFFER)/sizeof(EXPECTED_BUFFER[0]);

    //assert size of serialized buffer
    ASSERT_EQ(28, numBits);
    ASSERT_EQ(EXPECTED_BUFFER_SIZE, numBytes);

    //assert serialized buffer content
    for(int i=0; i<EXPECTED_BUFFER_SIZE; i++)
    {
      const uint8_t EXPECTED_CHARACTER = EXPECTED_BUFFER[i];
      const uint8_t actualCharacter = buffer[i];
      ASSERT_EQ(EXPECTED_CHARACTER, actualCharacter);
    }
  }
  //--------------------------------------------------------------------------------------------------
  TEST_F(TestBitReader, testAdd3bits)
  {
    #ifndef USE_BITADDRESS_READ_WRITE
    BitWriter bitwriter;
    #else
    BitAddress bitwriter;
    #endif

    //define an input buffer
    static const uint8_t INPUT_BUFFER[] = {0xaa, 0xbb, 0xcc};
    static const int INPUT_BUFFER_SIZE = sizeof(INPUT_BUFFER)/sizeof(INPUT_BUFFER[0]);

    //define an output buffer
    static const size_t MAX_BUFFER_SIZE = 30;
    uint8_t buffer[MAX_BUFFER_SIZE] = {0};

    //setup BitWriter
    bitwriter.setBuffer(buffer);

    //serialize
    for(int i=0; i<INPUT_BUFFER_SIZE; i++)
    {
      bitwriter.add(3);
      bitwriter.write(&INPUT_BUFFER[i], 8);
    }

    //compute total size of serialized data
    uint32_t numBits = bitwriter.toAbsBitOffset();
    uint32_t numBytes = numBits/8;
    if (numBits%8)
      numBytes++;

    //define ecpected buffer
    static const uint8_t EXPECTED_BUFFER[] = {0x50, 0xc5, 0x2e, 0x98, 0x01}; //expected serialized buffer content
    static const int EXPECTED_BUFFER_SIZE = sizeof(EXPECTED_BUFFER)/sizeof(EXPECTED_BUFFER[0]);

    //assert size of serialized buffer
    ASSERT_EQ(INPUT_BUFFER_SIZE*(8+3), numBits);
    ASSERT_EQ(EXPECTED_BUFFER_SIZE, numBytes);

    //assert serialized buffer content
    for(int i=0; i<EXPECTED_BUFFER_SIZE; i++)
    {
      const uint8_t EXPECTED_CHARACTER = EXPECTED_BUFFER[i];
      const uint8_t actualCharacter = buffer[i];
      ASSERT_EQ(EXPECTED_CHARACTER, actualCharacter);
    }
  }
  //--------------------------------------------------------------------------------------------------
  TEST_F(TestBitReader, testNext)
  {
    #ifndef USE_BITADDRESS_READ_WRITE
    BitWriter bitwriter;
    #else
    BitAddress bitwriter;
    #endif

    //define an input buffer
    static const uint8_t INPUT_BUFFER[] = {0xaa, 0xbb, 0xcc};
    static const int INPUT_BUFFER_SIZE = sizeof(INPUT_BUFFER)/sizeof(INPUT_BUFFER[0]);

    //define an output buffer
    static const size_t MAX_BUFFER_SIZE = 30;
    uint8_t buffer[MAX_BUFFER_SIZE] = {0};

    //setup BitWriter
    bitwriter.setBuffer(buffer);

    //serialize
    for(int i=0; i<INPUT_BUFFER_SIZE; i++)
    {
      bitwriter.next();
      bitwriter.next();
      bitwriter.next(); //same as add(3);

      bitwriter.write(&INPUT_BUFFER[i], 8);
    }

    //compute total size of serialized data
    uint32_t numBits = bitwriter.toAbsBitOffset();
    uint32_t numBytes = numBits/8;
    if (numBits%8)
      numBytes++;

    //define ecpected buffer
    static const uint8_t EXPECTED_BUFFER[] = {0x50, 0xc5, 0x2e, 0x98, 0x01}; //expected serialized buffer content
    static const int EXPECTED_BUFFER_SIZE = sizeof(EXPECTED_BUFFER)/sizeof(EXPECTED_BUFFER[0]);

    //assert size of serialized buffer
    ASSERT_EQ(INPUT_BUFFER_SIZE*(8+3), numBits);
    ASSERT_EQ(EXPECTED_BUFFER_SIZE, numBytes);

    //assert serialized buffer content
    for(int i=0; i<EXPECTED_BUFFER_SIZE; i++)
    {
      const uint8_t EXPECTED_CHARACTER = EXPECTED_BUFFER[i];
      const uint8_t actualCharacter = buffer[i];
      ASSERT_EQ(EXPECTED_CHARACTER, actualCharacter);
    }
  }
  //--------------------------------------------------------------------------------------------------
  TEST_F(TestBitReader, testDeserialize)
  {
    #ifndef USE_BITADDRESS_READ_WRITE
    BitReader bitreader;
    #else
    BitAddress bitreader;
    #endif

    static const uint8_t buffer[] = {0x41, 0x37, 0xfd, 0x9d, 0x76, 0x97, 0x01, 0x00, 0x00, 0x46, 0xa9, 0x02};
    static const int BUFFER_SIZE = sizeof(buffer)/sizeof(buffer[0]);

    //define an output Person
    Person p = {0};

    //setup BitWriter
    bitreader.setBuffer(buffer);

    //deserialize buffer into Person p
    for(int i=0; i<PERSON_NAME_SIZE; i++)
    {
      char c = 0;
      bitreader.read(7, &c);
      p.name[i] = c;
    }
    bitreader.read(5, &p.birthDay);
    bitreader.read(4, &p.birthMonth);
    bitreader.read(8, &p.birthYear);    p.birthYear += 1900;
    bitreader.read(1, &p.maried);
    bitreader.read(3, &p.children);

    //assert content properly extracted
    ASSERT_EQ(std::string("Antoine"), p.name);
    ASSERT_EQ(24, p.birthDay);
    ASSERT_EQ(8, p.birthMonth);
    ASSERT_EQ(1982, p.birthYear);
    ASSERT_EQ(true, p.maried);
    ASSERT_EQ(2, p.children);


  }
  //--------------------------------------------------------------------------------------------------
  TEST_F(TestBitReader, testSerialize)
  {
    #ifndef USE_BITADDRESS_READ_WRITE
    BitWriter bitwriter;
    #else
    BitAddress bitwriter;
    #endif

    //fill Person information
    Person p = {0};
    strcpy(p.name, "Antoine");
    p.birthDay = 24;
    p.birthMonth = 8;
    p.birthYear = 1982;
    p.maried = true;
    p.children = 2;

    //define an output buffer
    static const size_t MAX_BUFFER_SIZE = 100;
    uint8_t buffer[MAX_BUFFER_SIZE] = {0};

    //setup BitWriter
    bitwriter.setBuffer(buffer);

    //serialize Person p into buffer
    for(int i=0; i<PERSON_NAME_SIZE; i++)
    {
      char c = p.name[i];
      c = c & 0x7F;
      bitwriter.write(&c, 7);
    }
    bitwriter.write(&p.birthDay, 5);
    bitwriter.write(&p.birthMonth, 4);
    uint16_t tmpBirthYear = p.birthYear - 1900;
    bitwriter.write(&tmpBirthYear, 8);
    bitwriter.write(&p.maried, 1);
    bitwriter.write(&p.children, 3);

    //compute total size of serialized data
    uint32_t numBits = bitwriter.toAbsBitOffset();
    uint32_t numBytes = numBits/8;
    if (numBits%8)
      numBytes++;

    static const uint8_t EXPECTED_BUFFER[] = {0x41, 0x37, 0xfd, 0x9d, 0x76, 0x97, 0x01, 0x00, 0x00, 0x46, 0xa9, 0x02}; //expected serialized buffer content
    static const int EXPECTED_BUFFER_SIZE = sizeof(EXPECTED_BUFFER)/sizeof(EXPECTED_BUFFER[0]);

    //assert size of serialized buffer
    ASSERT_EQ(91, numBits);
    ASSERT_EQ(EXPECTED_BUFFER_SIZE, numBytes);

    //assert serialized buffer content
    for(int i=0; i<EXPECTED_BUFFER_SIZE; i++)
    {
      const uint8_t EXPECTED_CHARACTER = EXPECTED_BUFFER[i];
      const uint8_t actualCharacter = buffer[i];
      ASSERT_EQ(EXPECTED_CHARACTER, actualCharacter);
    }
  }
  //--------------------------------------------------------------------------------------------------
} // End namespace test
} // End namespace arduino


src/bitreader.h
// ---------------------------------------------------------------------------
// AUTHOR/LICENSE:
//  The following code was written by Antoine Beauchamp. For other authors, see AUTHORS file.
//  The code & updates for the library can be found at http://github.com/end2endzone/BitReader
//  MIT License: http://www.opensource.org/licenses/mit-license.php
// ---------------------------------------------------------------------------

#ifndef BITREADER_H
#define BITREADER_H

#include "Arduino.h"

#ifndef BITREADER_API
#define BITREADER_API
#endif

#ifndef BITS_PER_BYTE
#define BITS_PER_BYTE 8
#endif

//Bit optimization switches
#define USE_BITADDRESS_SETTER_GETTER
//#define USE_SINGLEBIT_MACROS
#define USE_BITADDRESS_READ_WRITE

struct BITREADER_API BitAddress
{
  unsigned int byte; //byte index
  unsigned char bit; //bit index

  BitAddress();
  BitAddress(const BitAddress & iBitAddress);
  void reset();
#ifdef USE_BITADDRESS_SETTER_GETTER
  void set(void * iBuffer, bool iBitValue);
  bool get(const void * iBuffer);
#endif
  void add(unsigned int iBits);
  void next();
  unsigned int toAbsBitOffset();

#ifdef WIN32
  std::string toString();
#endif

#ifdef USE_BITADDRESS_READ_WRITE
  void setBuffer(const void * iBuffer);
  void read(unsigned int iNumBits, void * iTargetBits);
  void write(const void * iSourceBits, unsigned int iNumBits);
  const unsigned char * mBuffer;
#endif

};

#ifndef USE_BITADDRESS_READ_WRITE

class BITREADER_API BitReader
{
public:
  BitReader();
  void setBuffer(const void * iBuffer);
  void read(unsigned int iNumBits, void * iTargetBits);
  unsigned char read8(unsigned int iNumBits);
  unsigned short read16(unsigned int iNumBits);
  const BitAddress & getAddress();
private:
  const unsigned char * mBuffer;
  BitAddress mAddr;
};

class BITREADER_API BitWriter
{
public:
  BitWriter();
  void setBuffer(void * iBuffer);
  void write(const void * iSourceBits, unsigned int iNumBits);
  void write8(unsigned char iSourceBits, unsigned int iNumBits);
  void write16(unsigned short iSourceBits, unsigned int iNumBits);
  void skip(unsigned int iNumBits);
  const BitAddress & getAddress();
private:
  unsigned char * mBuffer;
  BitAddress mAddr;
};

#endif //USE_BITADDRESS_READ_WRITE
#endif //BITREADER_H


src/BitReader_unittest/TestBitReader.cpp
#include <stdio.h>
#include <tchar.h>

#include <string>
#include "TestBitReader.h"
#include "arduino.h"
#include "BitReader.h"
#include "gTestHelper.h"

using namespace testarduino;

namespace arduino { namespace test
{
  static const size_t PERSON_NAME_SIZE = 10;

  struct Person {
    char name[PERSON_NAME_SIZE];  //10*7 bits, 70 bits
    uint8_t birthDay;             //from 1 to 31, 5 bits
    uint8_t birthMonth;           //from 1 to 12, 4 bits
    uint16_t birthYear;           //from 1900 to 2149, 250 years, 8 bits
    bool maried;                  //1 bit
    uint8_t children;             //from 0 to 7, 3 bits
  };

  //--------------------------------------------------------------------------------------------------
  void TestBitReader::SetUp()
  {
    //disable function logging
    testarduino::setLogFile("");
  }
  //--------------------------------------------------------------------------------------------------
  void TestBitReader::TearDown()
  {
  }
  //--------------------------------------------------------------------------------------------------
  TEST_F(TestBitReader, testDeserialize)
  {
    #ifndef USE_BITADDRESS_READ_WRITE
    BitReader bitreader;
    #else
    BitAddress bitreader;
    #endif

    static const uint8_t buffer[] = {0x41, 0x37, 0xfd, 0x9d, 0x76, 0x97, 0x01, 0x7e, 0x3f, 0x46, 0xa9, 0xca};
    static const uint8_t BUFFER_SIZE = sizeof(buffer)/sizeof(buffer[0]);

    //define an output Person
    Person p = {0};

    //setup BitWriter
    bitreader.setBuffer(buffer);

    //deserialize buffer into Person p
    for(int i=0; i<PERSON_NAME_SIZE; i++)
    {
      char c = 0;
      bitreader.read(7, &c);
      p.name[i] = c;
    }
    bitreader.read(5, &p.birthDay);
    bitreader.read(4, &p.birthMonth);
    bitreader.read(8, &p.birthYear);    p.birthYear += 1900;
    bitreader.read(1, &p.maried);
    bitreader.read(3, &p.children);

    //assert content properly extracted
    ASSERT_EQ(std::string("Antoine"), p.name);
    ASSERT_EQ(24, p.birthDay);
    ASSERT_EQ(8, p.birthMonth);
    ASSERT_EQ(1982, p.birthYear);
    ASSERT_EQ(true, p.maried);
    ASSERT_EQ(2, p.children);


  }
  //--------------------------------------------------------------------------------------------------
  TEST_F(TestBitReader, testSerialize)
  {
    #ifndef USE_BITADDRESS_READ_WRITE
    BitWriter bitwriter;
    #else
    BitAddress bitwriter;
    #endif

    //fill Person information
    Person p;
    strcpy_s(p.name, PERSON_NAME_SIZE, "Antoine");
    p.birthDay = 24;
    p.birthMonth = 8;
    p.birthYear = 1982;
    p.maried = true;
    p.children = 2;

    //define an output buffer
    static const size_t MAX_BUFFER_SIZE = 100;
    uint8_t buffer[MAX_BUFFER_SIZE];

    //setup BitWriter
    bitwriter.setBuffer(buffer);

    //serialize Person p into buffer
    for(int i=0; i<PERSON_NAME_SIZE; i++)
    {
      char c = p.name[i];
      c = c & 0x7F;
      bitwriter.write(&c, 7);
    }
    bitwriter.write(&p.birthDay, 5);
    bitwriter.write(&p.birthMonth, 4);
    uint16_t tmpBirthYear = p.birthYear - 1900;
    bitwriter.write(&tmpBirthYear, 8);
    bitwriter.write(&p.maried, 1);
    bitwriter.write(&p.children, 3);

    //compute total size of serialized data
    uint32_t numBits = bitwriter.toAbsBitOffset();
    uint32_t numBytes = numBits/8;
    if (numBits%8)
      numBytes++;

    static const uint8_t EXPECTED_BUFFER[] = {0x41, 0x37, 0xfd, 0x9d, 0x76, 0x97, 0x01, 0x7e, 0x3f, 0x46, 0xa9, 0xca}; //expected serialized buffer content
    static const uint8_t EXPECTED_BUFFER_SIZE = sizeof(EXPECTED_BUFFER)/sizeof(EXPECTED_BUFFER[0]);

    //assert size of serialized buffer
    ASSERT_EQ(91, numBits);
    ASSERT_EQ(EXPECTED_BUFFER_SIZE, numBytes);

    //assert serialized buffer content
    for(int i=0; i<EXPECTED_BUFFER_SIZE; i++)
    {
      const uint8_t EXPECTED_CHARACTER = EXPECTED_BUFFER[i];
      const uint8_t actualCharacter = buffer[i];
      ASSERT_EQ(EXPECTED_CHARACTER, actualCharacter);
    }
  }
  //--------------------------------------------------------------------------------------------------
} // End namespace test
} // End namespace arduino
